### PR TITLE
sql, server: background job for temp data cleanup

### PR DESCRIPTION
**Pull request by solongordon**
Refs: #45666

Normally, temporary objects are deleted when the session that created
them exits. However, it is possible for this hook not to get run if a
session does not exit gracefully, for example if the gateway node dies.
This causes the corresponding temp data to hang around indefinitely.

This commit adds a background job which cleans up such data. It
identifies and drops any temporary schemas which do not belong to an
active session. The job runs on an semi-hourly basis as long as it's
not already running on another node.

Release note: None

### CODE REVIEW

**Comment by solongordon** [comment]
@spaskob Would you mind just sanity checking the job-related stuff?

**Review by otan — CHANGES_REQUESTED** [review]
reviewed logic; not so much the job stuff as less familiar but it looks correct.

i think there is a race tho and we should probably change it.

**Review by spaskob — COMMENTED** [review]
LGTM for jobs related stuff

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45185#-:-M0SJYF_7WpEGKXJ2BZ8:b-hdj5hp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @solongordon and @spaskob)

File: pkg/sql/delegate/show_jobs.go
Comment (otan) [inline pkg/sql/delegate/show_jobs.go]: needs `''` around %s, i think

File: pkg/sql/opt/exec/execbuilder/testdata/explain
Comment (otan) [inline pkg/sql/opt/exec/execbuilder/testdata/explain]: (and this to be updated)

**Review by solongordon — COMMENTED** [review]
Yeah... I was lazy and relied on CI to catch my error. Should be good now.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45185#-:-M0TqdlB3U9kKsn__6Xy:brrmeso)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/temporary_object_cleanup.go, line 66 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45185#-M0PIWnF0kWgnyZWfDYI-r1-66:-M0T1u4qBiokCB0u2GsW:b-sqov36) ([raw file](https://github.com/cockroachdb/cockroach/blob/a980432bca93a73d1572eb64a47c9d1d56f4207c/pkg/sql/temporary_object_cleanup.go#L66)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

is there a race here where sessions can appear between `ListSessions` and this call?
should we do the `	response, err := p.extendedEvalCtx.StatusServer.ListSessions(ctx, &req)` AFTER this schema query instead? more lookups, but that should be ok as it is async.
</blockquote></details>

Yes, great catch. Fixed.
___
*[pkg/sql/delegate/show_jobs.go, line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45185#-M0PIWnF0kWgnyZWfDYE-r1-31:-M0T3FH44OExFZQOKVjL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a980432bca93a73d1572eb64a47c9d1d56f4207c/pkg/sql/delegate/show_jobs.go#L31)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

maybe worth using `in` and `not in`
</blockquote></details>

Done.
___
*[pkg/sql/delegate/show_jobs.go, line 30 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45185#-M0PIWnF0kWgnyZWfDYE-r2-30:-M0TqagH1t6BvCd3tw81:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/da0611810546591db1cf59586106192ff6cc623b/pkg/sql/delegate/show_jobs.go#L30)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

needs `''` around %s, i think 
</blockquote></details>

Done.
___
*[pkg/sql/opt/exec/execbuilder/testdata/explain, line 131 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45185#-M0PIWnF0kWgnyZWfDYG-r2-131:-M0TqbNn1naDHNl89_cq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/da0611810546591db1cf59586106192ff6cc623b/pkg/sql/opt/exec/execbuilder/testdata/explain#L131)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

(and this to be updated)
</blockquote></details>

Done.

**Review by otan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45185)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @solongordon)

___
*[pkg/sql/temporary_object_cleanup.go, line 88 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45185#-M0XBwg6CD8ugfn5bj7j:-M0XBwg6CD8ugfn5bj7k:b61fw1k) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a2bb64cb402e62f7963a9897fcdb25a154ae60e/pkg/sql/temporary_object_cleanup.go#L88)):*
> ```Go
> 	for sessionID := range sessionIDs {
> 		if _, ok := activeSessions[sessionID.Uint128]; !ok {
> 			ie := r.job.MakeSessionBoundInternalExecutor(ctx, sessionDataForCleanup(sessionID))
> ```

hmm, will this ever conflict with a job that is running at the same time?
my hunch is no as `activeSessions` will still be in place.
___
*[pkg/sql/temporary_object_cleanup.go, line 91 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45185#-M0XC-Jp4sUZI7bLMwMC:-M0XC-Jp4sUZI7bLMwMD:bw65w5d) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a2bb64cb402e62f7963a9897fcdb25a154ae60e/pkg/sql/temporary_object_cleanup.go#L91)):*
> ```Go
> 			if err := cleanupSessionTempObjects(
> 				ctx, p.execCfg.Settings, r.job.DB(), ie.Exec, sessionID); err != nil {
> 				return err
> ```

should we ignore errors and continue trying to delete other temp tables?
or should we block permanently if the one job in particular has failed? ok either way.
___
*[pkg/sql/temporary_object_cleanup.go, line 157 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45185#-M0XCYvKBBxCg-vhVJba:-M0XCYvKBBxCg-vhVJbb:b-o23xix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a2bb64cb402e62f7963a9897fcdb25a154ae60e/pkg/sql/temporary_object_cleanup.go#L157)):*
> ```Go
> 			); delErr != nil {
> 				log.Warningf(ctx, "failed to delete job: %v", delErr)
> 			}
> ```

do we wanna return `nil` after this?

**Review by otan — APPROVED** [review]
(lgtm after lint fix, below comments optional) 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45185#-:-M0XF6LnAKay2_AWxs0w:b-dzmov9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @solongordon)

**Comment by jordanlewis** [comment]
@otan, since Solon is now on leave, will you please take this PR over? Looks like there's not much work left to finish it.

**Comment by otan** [comment]
yes, i'll have a squiz at wrapping this up later tonight. just gotta fix a few commentz

**Comment by solongordon** [comment]
Thanks @otan! The unit test I added is also failing testrace unfortunately, so we should make sure that is resolved before trying to merge. You should be able to reproduce locally like so:
```
make stressrace PKG=./pkg/sql TESTS=TestTempTablesDeletionJob
```

**Review by otan — COMMENTED** [review]
found the stress condition, changed the timeout override to be channel based instead.
fixed the lint errors and rebased.
also renamed "temp tables" to "temp objects" for job deletion
@jordanlewis do you mind giving this a review?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45185#-:-M0vp7QoBSLscE4Vtcc_:b-qpm6wk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @solongordon)

___
*[pkg/sql/temporary_object_cleanup.go, line 91 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45185#-M0XC-Jp4sUZI7bLMwMC:-M0v8VLMCQL4WvDZH8oK:b-du2gqy) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a2bb64cb402e62f7963a9897fcdb25a154ae60e/pkg/sql/temporary_object_cleanup.go#L91)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

should we ignore errors and continue trying to delete other temp tables?
or should we block permanently if the one job in particular has failed? ok either way.
</blockquote></details>

going to let this return error and retry (from the top). we randomise the order of deleting sessionIDs anyway. we can always re-evaluate this later.
___
*[pkg/sql/temporary_object_cleanup.go, line 157 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45185#-M0XCYvKBBxCg-vhVJba:-M0vq-yl7gE7uPjXiCo1:b-qsseby) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a2bb64cb402e62f7963a9897fcdb25a154ae60e/pkg/sql/temporary_object_cleanup.go#L157)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

do we wanna return `nil` after this?
</blockquote></details>

fixed!

**Review by jordanlewis — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45185)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @solongordon)

___
*[pkg/sql/create_stats.go, line 420 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45185#-M17iBbcCJVMeGPMH5Oa:-M17iBbd5P0Jyp-qkb_P:b-9gcys4) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e80ba7a35f5fcfd5f6ddf5d823ce5ae2477fc00/pkg/sql/create_stats.go#L420)):*
> ```Go
> // just checks if there are any pending, running, or paused CreateStats jobs.
> func checkRunningStatsJobs(ctx context.Context, job *jobs.Job, p *planner) error {
> 	err := checkRunningJobs(ctx, job, p, jobspb.TypeCreateStats, jobspb.TypeAutoCreateStats)
> ```

nit: varargs allocates on every invocation (pretty sure). Best practice here would be to stop using varargs, and instead pass in a constant slice (well, package var constant) of the relevant types.

I don't know how frequently this is meant to be called though, if it's really rare I guess it doesn't really matter.
___
*[pkg/sql/exec_util.go, line 624 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45185#-M17igrh8FARc4IkWbM5:-M17igrh8FARc4IkWbM6:bctehdo) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e80ba7a35f5fcfd5f6ddf5d823ce5ae2477fc00/pkg/sql/exec_util.go#L624)):*
> ```Go
> 	DisableTempObjectsCleanupOnSessionExit bool
> 
> 	// TempObjectsCleanupCh forces a cleanup run for deleting all temporary objects.
> ```

This comment doesn't make that much sense to me - how does a channel force anything? And what's the `time.Time` that goes into the channel?

seems like it could just be chan struct{} if you don't need to actually send anything over it
___
*[pkg/sql/temporary_object_cleanup.go, line 91 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45185#-M0XC-Jp4sUZI7bLMwMC:-M17kaJiCgqZVxxsFexN:bzeivea) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a2bb64cb402e62f7963a9897fcdb25a154ae60e/pkg/sql/temporary_object_cleanup.go#L91)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

going to let this return error and retry (from the top). we randomise the order of deleting sessionIDs anyway. we can always re-evaluate this later.
</blockquote></details>

I think we should ignore errors and continue trying to delete other temp tables.
___
*[pkg/sql/temporary_object_cleanup.go, line 53 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45185#-M17jYAM9OcxrRKRAcMv:-M17jYAM9OcxrRKRAcMw:b-4edaar) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e80ba7a35f5fcfd5f6ddf5d823ce5ae2477fc00/pkg/sql/temporary_object_cleanup.go#L53)):*
> ```Go
> 
> 	// Build a set of all session IDs with temporary objects.
> 	dbIDs, err := GetAllDatabaseDescriptorIDs(ctx, p.txn)
> ```

This might plausibly fail with a retry error... in which case, doesn't this only try again 30 minutes later? I feel like that's no good and we might need to run this stuff in a retry loop...
___
*[pkg/sql/temporary_object_cleanup.go, line 88 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45185#-M17k0xc5m4n1SmDd6dg:-M17k0xc5m4n1SmDd6dh:b-jy4f3r) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e80ba7a35f5fcfd5f6ddf5d823ce5ae2477fc00/pkg/sql/temporary_object_cleanup.go#L88)):*
> ```Go
> 	for sessionID := range sessionIDs {
> 		if _, ok := activeSessions[sessionID.Uint128]; !ok {
> 			ie := r.job.MakeSessionBoundInternalExecutor(ctx, sessionDataForCleanup(sessionID))
> ```

one executor per inactive session seems expensive - i think an internal executor isn't exactly a cheap structure.
___
*[pkg/sql/temporary_object_cleanup.go, line 180 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45185#-M17kqky6pFGZMvDrAwe:-M17kqky6pFGZMvDrAwf:bfuiabt) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e80ba7a35f5fcfd5f6ddf5d823ce5ae2477fc00/pkg/sql/temporary_object_cleanup.go#L180)):*
> ```Go
> 			select {
> 			case <-ch:
> 				if err := stopper.RunTask(ctx, "maybeDeleteTemporaryObjectss", func(ctx context.Context) {
> ```

does this task have a lisp?
___
*[pkg/sql/temporary_object_cleanup.go, line 185 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45185#-M17mB_b3ZfNE-qoFCov:-M17mB_b3ZfNE-qoFCow:bju7a31) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e80ba7a35f5fcfd5f6ddf5d823ce5ae2477fc00/pkg/sql/temporary_object_cleanup.go#L185)):*
> ```Go
> 					defer cleanup()
> 					if err := startTempObjectsCleanupJob(ctx, p); err != nil {
> 						log.Errorf(ctx, "failed to delete temp tables: %v", err)
> ```

Isn't this going to cause a ton of log spam? Since on n-1/n nodes won't this fail every time with the concurrent job error?
___
*[pkg/sql/temporary_schema.go, line 68 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45185#-M17l6UDDQzGxyLY3NRq:-M17l6UDDQzGxyLY3NRr:b6fsgur) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e80ba7a35f5fcfd5f6ddf5d823ce5ae2477fc00/pkg/sql/temporary_schema.go#L68)):*
> ```Go
> // temporarySchemaSessionID returns the sessionID of the given temporary schema.
> func temporarySchemaSessionID(scName string) (bool, ClusterWideID, error) {
> 	if !strings.HasPrefix(scName, "pg_temp_") {
> ```

nit: i bet we now have `pg_temp` scattered around in many strings, should we pull it into a constant?

**Comment by dt** [comment]
I'm curious but didn't see a mention in the PR desc: in the happy case sessions cleanup after themselves so these jobs are all no-ops, right? Do we we need to even create them?
i.e. would it make sense to scan for orphaned tables before creating a job and skip it if not?

**Review by otan — COMMENTED** [review]
> in the happy case sessions cleanup after themselves so these jobs are all no-ops, right? 
correct

> Do we we need to even create a persistent, resumable job if just to check if there is work to do?
I think this is an "optimization" of making sure we don't check all the tables more times than necessary.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45185#-:-M17xWUJ7Fp534e1Nnfb:b-8bu257)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @solongordon)

___
*[pkg/sql/create_stats.go, line 420 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45185#-M17iBbcCJVMeGPMH5Oa:-M17wY1C370a6wEy0X7n:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e80ba7a35f5fcfd5f6ddf5d823ce5ae2477fc00/pkg/sql/create_stats.go#L420)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

nit: varargs allocates on every invocation (pretty sure). Best practice here would be to stop using varargs, and instead pass in a constant slice (well, package var constant) of the relevant types.

I don't know how frequently this is meant to be called though, if it's really rare I guess it doesn't really matter.
</blockquote></details>

Done.
___
*[pkg/sql/exec_util.go, line 624 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45185#-M17igrh8FARc4IkWbM5:-M17otPO4enenXvl3YzE:braedoc) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e80ba7a35f5fcfd5f6ddf5d823ce5ae2477fc00/pkg/sql/exec_util.go#L624)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This comment doesn't make that much sense to me - how does a channel force anything? And what's the `time.Time` that goes into the channel?

seems like it could just be chan struct{} if you don't need to actually send anything over it
</blockquote></details>

usually we use a `time.Ticker.C` (see SetupBackgroundTempObjectsDeletionJob), which outputs `time.Time`, and making that ticker.C send a struct{} to another channel makes this a little more annoying. 
I've updated the comment
___
*[pkg/sql/temporary_object_cleanup.go, line 91 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45185#-M0XC-Jp4sUZI7bLMwMC:-M17on6k-pvSWVWY8fpJ:bw10k1l) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a2bb64cb402e62f7963a9897fcdb25a154ae60e/pkg/sql/temporary_object_cleanup.go#L91)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I think we should ignore errors and continue trying to delete other temp tables.
</blockquote></details>

Okay, change made accordingly. 
___
*[pkg/sql/temporary_object_cleanup.go, line 53 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45185#-M17jYAM9OcxrRKRAcMv:-M17yB2I3Hgox3_vbZ74:b70jmh4) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e80ba7a35f5fcfd5f6ddf5d823ce5ae2477fc00/pkg/sql/temporary_object_cleanup.go#L53)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This might plausibly fail with a retry error... in which case, doesn't this only try again 30 minutes later? I feel like that's no good and we might need to run this stuff in a retry loop...
</blockquote></details>

does this apply to getSchemasForDatabase too?
do we have a retry library? i don't want to rewrite one, hehe.
___
*[pkg/sql/temporary_object_cleanup.go, line 88 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45185#-M17k0xc5m4n1SmDd6dg:-M17stFfCtEmyn6VI5WE:bun6d63) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e80ba7a35f5fcfd5f6ddf5d823ce5ae2477fc00/pkg/sql/temporary_object_cleanup.go#L88)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

one executor per inactive session seems expensive - i think an internal executor isn't exactly a cheap structure.
</blockquote></details>

Fixed by adding a new `SetSessionData` onto the `ie`. We need the SearchPath of this updated so that the TableResolver doesn't error as it thinks your temporary schema does not belong to your session.
___
*[pkg/sql/temporary_object_cleanup.go, line 180 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45185#-M17kqky6pFGZMvDrAwe:-M17o3lv1VrgwNkJS-_J:b-ck3vhm) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e80ba7a35f5fcfd5f6ddf5d823ce5ae2477fc00/pkg/sql/temporary_object_cleanup.go#L180)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

does this task have a lisp?
</blockquote></details>

it's a snake! fixed.
___
*[pkg/sql/temporary_object_cleanup.go, line 185 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45185#-M17mB_b3ZfNE-qoFCov:-M17oUREDv1va_mVoeVt:b-q2cprv) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e80ba7a35f5fcfd5f6ddf5d823ce5ae2477fc00/pkg/sql/temporary_object_cleanup.go#L185)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Isn't this going to cause a ton of log spam? Since on n-1/n nodes won't this fail every time with the concurrent job error?
</blockquote></details>

the channel `<-ch` is hooked to a `ticker.C` of every 30mins. Is that too spammy?

**Comment by spaskob** [comment]
Hi, I'd like to take one more look today at this PR before this is merged please don't merge before my comments come in.

**Comment by ajwerner** [comment]
It seems like maybe we shouldn't use the jobs infrastructure here at all. One thing jobs in theory provide that can be useful is a claim of mutual exclusion. The exact contract of this mutual exclusion isn't super well specified but let's just assume that it provides a guarantee of mutual exclusion. Do we need that here? It seems like no. This work feels like it could be accomplished with a background loop that runs periodically and has only best-effort mutual exclusion.

There's precedent to setting up a loop that runs on every node but only executes the logic on a single node. See https://github.com/cockroachdb/cockroach/blob/22544bfbc60ef39eb3721f4d1dded6f0ab43d5ec/pkg/storage/reports/reporter.go#L104 or https://github.com/cockroachdb/cockroach/blob/22544bfbc60ef39eb3721f4d1dded6f0ab43d5ec/pkg/storage/protectedts/ptreconcile/reconciler.go#L103

The advantage is that we won't be abusing a subsystem we know we dislike with a new use case - permanently running, cluster-wide background cleanup. The downside is that it's a bit more primitive but I think that's fine. All of these loops can be easily replaced in the future. The risk of adopting jobs is that we'll begin to depend on code we'd like to rework. Thoughts?

**Comment by dt** [comment]
Sorry about the late drive-by here but I only just saw this.

As is, this feels like a strange "job" to me. Unlike all our other jobs, the record it persists doesn't actually define what work it is going to do (and we expect it to be a no-op most of the time) and it doesn't persist any execution state or indicate progress to the user. Those are the defining characteristics of a job<sup>1</sup>, but are ignored here -- it don't need a durable record of what we wanted to do or what we'd done so far, since every node can just do the same thing at any time. The job is here only as means of mutual exclusion / to coordinate execution, if I understand correctly?

If so, like @ajwerner says above, I think we can skip the job entirely.

I _do_ think it could be nice to still create a job _when there is work to do_ though as a way to communicate to the user that we're doing work i.e. we when determine there are sessions or tables to cleanup we spawn a job to do that. To be the most jobs-y I think said job would contain the table IDs or session IDs it is supposed to cleanup and then would be able to record which are done in its persisted progress, but even with the current code that discovers what is to do and ignores progress, I think it'd feel a lot more job-y if it was only created when there was in fact work to do.

<sup>1</sup> The defining characteristic of jobs is that they are some big piece of work which requires we _persist_ the work definition and its execution state.  E.g. When a user or some other trigger determines that we want to drop table x, restore tables a and b from backup foo, recompute stats on table z, etc, we persist that as a job so it can be resumed if the node that started it crashes, or so we can control and inspect its execution, e.g. pause it, resume it from paused on a different node, see it in the admin UI, etc. Finally, since jobs are assumed to be big pieces of work/interesting events, jobs get more heavily logged and surfaced to users (e.g. the UI) so operators can see what work their cluster is doing.

**Comment by ajwerner** [comment]
> I do think it could be nice to still create a job when there is work to do though as a way to communicate to the user that we're doing work i.e. we when determine there are sessions or tables to cleanup we spawn a job to do that. To be the most jobs-y I think said job would contain the table IDs or session IDs it is supposed to cleanup and then would be able to record which are done in its persisted progress, but even with the current code that discovers what is to do and ignores progress, I think it'd feel a lot more job-y if it was only created when there was in fact work to do.

I think it's a little weird from a UX perspective to say, hey, a node crashed, so now we're going to have a user-visible background job, but normally we don't. It's fine, and even sort of nice from an observability perspective, but doesn't seem like a requirement. Some metrics and a little logging feels okay to me.

It seems like then either we should create a job to drop temp tables always or we should never create a job. Eventually we'd like to make temp tables cheaper, let's not get users in the mindset that they're a heavyweight resource to be cleaned up with a job. In short, I’m opposed to using jobs here at all.

**Comment by dt** [comment]
> It seems like then either we should create a job to drop temp tables always or we should never create a job.

Great point -- I think I now agree with your prior suggestion of maybe shouldn't use the jobs infrastructure here at all.

**Comment by otan** [comment]
So it sounds like in summary, we're okay for this just ticking over in the background in every node, cleaning up schemas but without creating a job since our abstraction of job seems to be:
* user facing
* idea that there is significant work that needs to be done
* we need persist the work definition and its execution stat
which this job doesn't fit.

Seems like we're arguing that there are no concerns about:
* monitorability - sounds like basic logging is good enough (i'm a little on the minus side on this - it feels important that users know this is happening and can take up non-trivial resources - that said we should make this cheaper in future [but i am unaware of the long term plans of this]
* multiple nodes executing this at the same time because the task itself should be idempotent

**Comment by ajwerner** [comment]
> monitorability - sounds like basic logging is good enough (i'm a little on the minus side on this - it feels important that users know this is happening and can take up non-trivial resources - that said we should make this cheaper in future [but i am unaware of the long term plans of this]

We should absolutely have metrics. Counters for running cleanups, completed ones, other counters for how much work was done etc. Yay metrics!

Here's an example of similar metrics for a similar thing:

https://github.com/cockroachdb/cockroach/blob/master/pkg/storage/protectedts/ptreconcile/metrics.go


>  just ticking over in the background in every node, cleaning up schemas

We should only actually attempt to do the cleanup on a single node in the cluster. The `meta1` leaseholder is the traditional choice. See the links I posted above.

**Comment by dt** [comment]
> Seems like we're arguing that there are no concerns about:

> monitorability ...  important that users know this is happening and can take up non-trivial resources

Wouldn't that also apply to closing a session with lots of / large temp tables normally, where the executor does the cleanup as expected? I think what @ajwerner and I are arguing for here is consistency -- either there's always a job for cleanup or there never is. Making a job just for cleanup of _orphaned_ temp tables doesn't seem to make sense.

> multiple nodes executing this at the same time because the task itself should be idempotent

I think @ajwerner had a suggestion for that above and pointed to a couple examples.

**Comment by otan** [comment]
Sorry lots of comments missed the examples.

No worries, of course counters too :) thanks for the pointers, I'll follow up with a non-joby version of this PR next week.

[the third changing of the hands!]

**Comment by otan** [comment]
I've taken these suggestions and put them in https://github.com/cockroachdb/cockroach/pull/45669 instead.

**Pull request closed without merging**
